pjsip/src/pjsip-ua/sip_inv.c

```diff
@@ -2664,8 +2664,8 @@ PJ_DEF(pj_status_t) pjsip_inv_initial_answer(   pjsip_inv_session *inv,
         pj_status_t status2;
 
         status2 = pjsip_dlg_modify_response(inv->dlg, tdata, st_code2, NULL);
+        pjsip_tx_data_dec_ref(tdata);
         if (status2 != PJ_SUCCESS) {
-            pjsip_tx_data_dec_ref(tdata);
```

Comment:
Here the ref count is not decremented when not-success, but below after `pjsip_timer_update_resp` it is decremented. Is it okay?

Also, please make sure that the pattern is consistent for every `pjsip_dlg_modify_response` in the lib.

Reply:
The issue is that the `tdata->ref_cnt` will be 2 when `pjsip_inv_initial_answer()` returns. 
e.g: the `pjsip_tx_data_dec_ref()` after `pjsip_timer_update_resp()` will not release the `tdata`.
I think setting the `tdata->last_answer`  would be better to avoid the inconsistency and let inv/tsx decrement it later.

Reply:
- Setting `inv->last_answer` is fine **only when** `p_tdata` is set (so app will send it or dec ref it), otherwise it will still leak?
- `pjsip_dlg_modify_response` does not mention about add ref count in its spec/doc, I think it's better to add it now.
- Also from a quick check, in some other places, e.g: when `process_answer()` fails, it seems that the tdata ref count is not properly decremented? (need test all affected scenarios to make sure things are fine).

Reply:
* At the moment we are missing a dec ref due to `pjsip_dlg_modify_response`. So, it will still leak even if `p_tdata` is not set. We can set `inv->last_answer` only if `p_tdata` is set, however we need to call an additional `pjsip_tx_data_dec_ref` in the case `p_tdata` is not set.
* Right, will add the info to the doc
* You are right, we need to check the code after `pjsip_dlg_modify_response`. We need to an additional `pjsip_tx_data_dec_ref`
e.g:
https://github.com/pjsip/pjproject/blob/43dd94f4ec257993959f381ea46091ece7aaf85d/pjsip/src/pjsip-ua/sip_inv.c#L2758

Reply:
I am actually a bit puzzled as to why we need to add tdata ref in `pjsip_dlg_modify_response()`. Is there a way for us to avoid adding the ref?

- Adding the reference in a function called 'modify' doesn't seem very intuitive, and rather unexpected. Even we did not know this before (and especially so because the doc never says anything about it).
- It also seems "ugly" if we have to decrement the ref twice on some failure cases.

From the code, it states: "Must add reference counter, since tsx_send_msg() will decrement it." And it will probably causes crash if we remove it, but perhaps there's another way?

Reply:
Well, the API has been there since the beginning, see [here](https://github.com/pjsip/pjproject/blob/999b896352d3ea00d2920940e44975c202aae23f/pjsip/src/pjsip/sip_dialog.c#L1353-L1354), so I think modifying the behavior will cause a major backward compatibility issue.

We can always introduce an alternative API (without add ref) and replace all usages of the old one in the library, but perhaps too much hassle and we may still need to maintain the old one for quite some time for existing apps.

Reply:
Yes, if it introduces backward incompatibility, I agree that we shouldn't change it. Also, we don't need to create additional API since it's not a major issue.

I was just wondering why we need the additional tdata ref increment inside `modify()`, and making the tdata ref to become two, while if we use `create()`, it seems that the tdata ref is only one before we send it?

Reply:
Just a guess. Perhaps because `modify_response()` is designated to work with a saved `tdata` such as `inv->last_answer` or `tsx->last_tx`, so the saved `tdata` will not be destroyed after being sent, e.g:
```
tdata = inv->last_answer;
modify_response(tdata);
send_response(tdata);
/* inv->last_answer will still be maintained here */
```